proposals/threads/Globals.md

```diff
@@ -2,6 +2,341 @@
 
 This page describes a proposal for importing and exporting mutable globals.
 
+## Rationale
+
+Without the ability to import and export mutable globals, it is inconvenient to
+provide mutable thread-local values that can be dynamically linked, such as the
+C++ stack pointer (SP).
+
+The following examples use SP as a motivating example, but a similar argument
+can be made for other thread-local variables (or the TLS pointer itself) as
+well.
+
+### Example: Dynamically Linking SP with Single Agent
+
+Let's assume we have two modules that are dynamically linked, `m1` and
+`m2`. They both use the C++ stack pointer. In the MVP, it's not possible to
+import or export a mutable global. In a single-agent program, we can work
+around this by storing the stack pointer in linear memory:
+
+```
+(module $m1
+  (memory (export "memory") 1)
+  ;; Address 0 is the address of SP
+  ;; The stack starts at 0x100
+  (data (i32.const 0) "\00\01\00\00)
+
+  (func
+    ;; SP = SP + 64
+    (i32.store
+      (i32.const 0)  ;; SP
+      (i32.add
+        (i32.load (i32.const 0))  ;; SP
+        (i32.const 64)))
+    ...
+  )
+  ...
+)
+
+(module $m2
+  (import "env" "memory" (memory 1))
+  ...
+)
+```
+
+Then the modules can be instantiated as follows:
+
+```
+WebAssembly.instantiate(m1Bytes, {}).then(
+    ({instance}) => {
+        let imports = {env: {memory: instance.exports.memory}};
+        WebAssembly.instantiate(m2Bytes, imports).then(...);
+    });
+```
+
+This solution does not work when linear memory is shared, since we need
+different memory locations for each stack pointer (one per agent).
+
+### Solution 1: Use Per-Module Immutable Global
+
+One solution is to use a per-module immutable global to store the location of
+the SP in linear memory:
+
+```
+(module $m1
+  (import "env" "spAddr" (global $spAddr i32))
+  (memory (export "memory") 1)
+
+  (data (i32.const 0) "\00\01\00\00")  ;; SP for agent 0
+  (data (i32.const 4) "\00\02\00\00")  ;; SP for agent 1
+
+  (func
+    ;; SP = SP + 64
+    (i32.store
+      (get_global $spAddr)
+      (i32.add
+        (i32.load (get_global $spAddr))
+        (i32.const 64)))
+    ...
+  )
+  ...
+)
+
+(module $m2
+  (import "env" "memory" (memory 1))
+  (import "env" "spAddr" (global $spAddr i32))
+  ...
+)
+```
+
+Then the modules can be instantiated as follows:
+
+```
+let agentIndex = 0;
+let spAddrs = [0x0, 0x4];
+let imports = {env: {spAddr: spAddrs[agentIndex]}};
+WebAssembly.instantiate(m1Bytes, imports).then(
+    ({instance}) => {
+        let imports = {env: {
+          memory: instance.exports.memory,
+          spAddr: spAddrs[agentIndex],
```

Comment:
A different `agentIndex` would be in another worker... so this example is a bit confusing. I'll clarify.